src/core/lib/security/security_connector/tls/tls_security_connector.cc

```diff
@@ -138,54 +74,166 @@ grpc_error* TlsCheckHostName(const char* peer_name, const tsi_peer* peer) {
+            "TlsChannelSecurityConnectorCreate()");
+    return nullptr;
+  }
+  grpc_core::RefCountedPtr<TlsChannelSecurityConnector> c =
+      grpc_core::MakeRefCounted<TlsChannelSecurityConnector>(
+          std::move(ch_creds), std::move(request_metadata_creds), target_name,
+          overridden_target_name, ssl_session_cache);
+  // The session resumption information is invalidated after we create the
+  // security connector instance. It will only be used if we invoke
+  // OnCertificatesChanged when creating security connector.
+  c->ssl_session_cache_ = nullptr;
+  return c;
+}
+
 TlsChannelSecurityConnector::TlsChannelSecurityConnector(
-    grpc_core::RefCountedPtr<grpc_channel_credentials> channel_creds,
+    grpc_core::RefCountedPtr<grpc_channel_credentials> ch_creds,
     grpc_core::RefCountedPtr<grpc_call_credentials> request_metadata_creds,
-    const char* target_name, const char* overridden_target_name)
-    : grpc_channel_security_connector(GRPC_SSL_URL_SCHEME,
-                                      std::move(channel_creds),
+    const char* target_name, const char* overridden_target_name,
+    tsi_ssl_session_cache* ssl_session_cache)
+    : grpc_channel_security_connector(GRPC_SSL_URL_SCHEME, std::move(ch_creds),
                                       std::move(request_metadata_creds)),
       overridden_target_name_(
-          overridden_target_name == nullptr ? "" : overridden_target_name) {
-  key_materials_config_ = grpc_tls_key_materials_config_create()->Ref();
+          overridden_target_name == nullptr ? "" : overridden_target_name),
+      ssl_session_cache_(ssl_session_cache) {
   check_arg_ = ServerAuthorizationCheckArgCreate(this);
   absl::string_view host;
   absl::string_view port;
   grpc_core::SplitHostPort(target_name, &host, &port);
   target_name_ = std::string(host);
+  // Create a watcher.
+  std::unique_ptr<
+      grpc_tls_certificate_distributor::TlsCertificatesWatcherInterface>
+      watcher_ptr = absl::make_unique<TlsChannelCertificateWatcher>(this);
+  certificate_watcher_set_.insert(watcher_ptr.get());
+  // Register the watcher with the provider.
+  const TlsCredentials* creds =
+      static_cast<const TlsCredentials*>(channel_creds());
+  grpc_tls_certificate_provider* provider =
+      creds->options()->certificate_provider();
+  GPR_ASSERT(provider != nullptr);
+  RefCountedPtr<grpc_tls_certificate_distributor> distributor =
+      provider->distributor();
+  distributor->WatchTlsCertificates(std::move(watcher_ptr),
+                                    creds->options()->root_cert_name(),
+                                    creds->options()->identity_cert_name());
 }
 
 TlsChannelSecurityConnector::~TlsChannelSecurityConnector() {
+  gpr_log(GPR_ERROR, "TlsChannelSecurityConnector dtor is called");
+  // Cancel all the watchers.
+  const TlsCredentials* creds =
+      static_cast<const TlsCredentials*>(channel_creds());
+  grpc_tls_certificate_provider* provider =
```

Comment:
Hi @markdroth , I have another question here: can we always assume the provider will outlive the security connector? At the time when the security connector got destroyed, if the provider is already gone, we are not able to get the provider here and unregister the watcher. 
It might make sense for the provider to notify the security connector when it is destroyed, but I just want to see if the above assumption makes sense here.